Rewrite gas mechanisms in terms of explicit gas_charge nodes in the AST

vaivaswatha opened the pull request:
The AST is transformed to include explicit gas_charge nodes. The evaluator just evaluates each gas_charge node and does not do any separate gas charging.

This enables the compiler to just generate code for gas_charge expressions and achieve uniform gas charging in both the implementations.

There are minor gas charge differences arising out of this PR:

1. From commit 6e3d4a1a495fc6d3fc0c0c6ea1c8886ebc0763e8
    1. Places where we used String.length is now replaced with literal_cost of the string, which is computed differently.
    2. Places where length of "pp_literal l" was used is now replaced with literal_cost. This causes minor differences too.
    3. For the Builtin_alt_bn128_G1_mul, computation of Logarithm of argument now first adds one and then proceeds.

2. From commit 45b53bebdc09b658279c5401382e2ac5923b041f
    1. Due to computing the cost of `send` as the size of it's "Scilla List of Scilla Messages" (now) vs "OCaml List of Scilla Message" earlier.

jjcnn commented on src/base/Gas.ml:
```diff
@@ -175,204 +172,235 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
-        let x = div_ceil (Bystr.width s + 66) 64 in
-        pure @@ ((250 + (15 * x)) * base)
+        let s = GasCharge.SizeOf (GI.get_id a) in
+        let n = GasCharge.StaticCost (64 * 15) in
```
The previous gas charge was size * 64 * 10, but now it's size * 64 * 15. I don't know which one is correct.

jjcnn commented on src/base/Gas.ml:
```diff
@@ -104,70 +104,157 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+          | Iterate (l, _) ->
+              let g = GasStmt (GasCharge.LengthOf (GI.get_id l)) in
+              [ (g, srep); (s, srep) ]
+          | Throw _ (* TODO: Throw should charge same as event and send. *)
```
This TODO needs doing.

vaivaswatha replied:
I wanted to keep the gas changes (in tests) minimal, retaining as much as possible current gas charges. I'll do this as a separate change right after we are done with this PR (immediately, so I won't create an Issue to track it).

jjcnn commented on src/base/Gas.ml:
```diff
@@ -104,70 +104,157 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+              [ (g, srep); (s, srep) ]
+          | Throw _ (* TODO: Throw should charge same as event and send. *)
+          | GasStmt _ ->
+              [ (s, srep) ]
```
I think this should throw an error. The funtion adds gas charge statements before or after the statement for which gas is charged, so if there is already a gas charge statement in the AST, then something has gone wrong.

vaivaswatha replied:
There are two reasons I left it this way:

1. Handling errors here (monadically) has a domino effect up until Runner.ml, and that means I'll have to manually check for errors at multiple places there (making it a bit ugly in Runner.ml).
2. More importantly, I wanted to keep it this way so that in case we introduce explicit gas charge statements / expressions in the language (though I can't think of a reason for having such a thing right now), this will just work.

But having said that, I am flexible about this, and if you think it's better to do what you said, I'll do it.

jjcnn replied:
1. You don't need to handle errors here. You just need to raise an error. It would be an internal error because this branch shouldn't happen.
2. If we want to add explicit gas charges, then this branch should be changed, but until we do we would expect this to be an error. And at that point we will discover that this branch should be changed, because we will hit it. I don't see any reason to not throw an error here.

jjcnn commented on src/eval/Eval.ml:
```diff
@@ -113,6 +113,9 @@ let eval_gas_charge env g =
+    | GasCharge.MapSortCost vstr ->
+        let%bind m = Env.lookup env (mk_loc_id vstr) in
+        pure @@ EvalGas.map_sort_cost m
     | _ -> fail0 "eval_gas_charge: Must be handled by GasCharge"
```
Do you mind adding the missing GasCharge constructors explicitly in this pattern match? I don't like the idea that we add gas charge constructors and the OCaml compiler not complaining about a missing case.

jjcnn reviewed (changes requested):
I have added a couple of suggestions and commments.

Other than that I don't think it's ideal to have gas charge as a separate statement type, so that every second statement is a gas charge statement. I think the charging of gas should be tied more closely to the statement that incurs the gas charge.

My worry is that either compiler optimisations will be impossible (because gas charge statements prevent us from combining statements, e.g., through deforestation), or that gas charge statements are optimised out by accident (perhaps by the LLVM JIT compilation) because they don't affect the final result.

I don't really know how else to do it, though, but doing it this way makes my spine tingle in an upleasant way...

vaivaswatha commented on src/base/Gas.ml:
```diff
@@ -104,70 +104,157 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+                  GasCharge.SumOf
+                    ( GasCharge.SumOf
+                        ( GasCharge.SizeOf (GI.get_id x),
+                          GasCharge.MapSortCost (GI.get_id x) ),
```
Because when we fetch a Map value from the blockchain, the protobuf transmission doesn't guarantee any order on key-val pairs (i.e., maps). (Note that leveldb storage is ordered, so direct key accesses are still efficient). So to guarantee an order for Scilla, we sort map values when bringing them in. That's the cost.

jjcnn replied:
But why do we need to order the map on the Scilla side? Don't we just drop it all down a hashmap?

vaivaswatha replied:
After referring back to https://github.com/Zilliqa/scilla/pull/764, I recall that we had an issue with reaching consensus because of this. With a non-deterministic order, two different Scilla executions, on different nodes, can lead to different outputs.

jjcnn replied:
Oh, right - there was something like that.

It looks like Jacob in January knew what he was talking about, so who is Jacob in October to question that...

vaivaswatha commented on src/base/Gas.ml:
```diff
@@ -104,70 +104,157 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+              let n = GasCharge.StaticCost (List.length klist) in
+              let g =
+                match ropt with
+                | Some r -> GasCharge.SumOf (GasCharge.SizeOf (GI.get_id r), n)
```
I haven't implemented this suggestion. It isn't just map values we're fetching here, but could be any value. So it's plain size of value fetched (which probably doesn't warrant a comment, but I can add one if you think it's better after reading my response).

jjcnn replied:
Nah, never mind.

vaivaswatha commented on src/base/Gas.ml:
```diff
@@ -104,70 +104,157 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
+                GasStmt
+                  (GasCharge.SumOf
+                     ( GasCharge.SizeOf (GI.get_id x),
+                       GasCharge.MapSortCost (GI.get_id x) ))
```
Answered below.

vaivaswatha commented on src/base/Gas.ml:
```diff
@@ -175,204 +172,235 @@ module ScillaGas (SR : Rep) (ER : Rep) = struct
-        let x = div_ceil (Bystr.width s + 66) 64 in
-        pure @@ ((250 + (15 * x)) * base)
+        let s = GasCharge.SizeOf (GI.get_id a) in
+        let n = GasCharge.StaticCost (64 * 15) in
```
1. Fixed the typo / bug.
2. We do have tests for ~schnorr~ `Builtin_ripemd160hash`, but they are all as expression tests (not full contract tests). And at the moment we don't print gas remaining in eval-runner. I plan to push in a PR to do that after this.

vaivaswatha commented:
Thank you for the review Jacob.

> I have added a couple of suggestions and commments.

I've addressed the individual comments, either by making the changes or responding to your comment.

Regarding your take on the overall change:

> Other than that I don't think it's ideal to have gas charge as a separate statement type, so that every second statement is a gas 
charge statement. I think the charging of gas should be tied more closely to the statement that incurs the gas charge.

This was conceptually what was happening earlier too. The evaluator would generate a gas charge descriptor after each statement, and then later that was evaluated to charge gas. We are just doing that statically now.

> My worry is that either compiler optimisations will be impossible (because gas charge statements prevent us from combining statements, e.g., through deforestation), 

I agree that, to an extent, compiler optimizations may be impacted, but not made impossible. You can still, for example, combine statements. You'll have to be more smart about it. For example, do both computations first, and then both gas charges. If the gas charge is before a computation, do both charges before, and then both the computations (this is safe because, it would run out of gas anyway if it was supposed to, but now, without actually doing that computation).

>or that gas charge statements are optimised out by accident (perhaps by the LLVM JIT compilation) because they don't affect the final result.

No, this won't happen. Gas charge statements are (currently) compiled to be statements with side effect (check and subtract from a global "gas_remaining" variable). Even if compiled in other ways (say a more "pure" way), the final remaining gas will be a part of the final result, and hence not optimized away.

Thanks again.

jjcnn commented:
[snip]
> 
> > My worry is that either compiler optimisations will be impossible (because gas charge statements prevent us from combining statements, e.g., through deforestation),
> 
> I agree that, to an extent, compiler optimizations may be impacted, but not made impossible. You can still, for example, combine statements. You'll have to be more smart about it. For example, do both computations first, and then both gas charges. If the gas charge is before a computation, do both charges before, and then both the computations (this is safe because, it would run out of gas anyway if it was supposed to, but now, without actually doing that computation).

Except that with this change you don't know (for statements) whether a gas charge is connected to the previous or the next statement. We knew that before, so we could just optimise away without worrying about gas, and the gas charge would follow the optimised code.

Also, it's not just about whether we run out of gas or not. It's also about the amount of work the miner needs to do before we run out of gas. Say that your program is as follows (`s` for statement, `g` for gas charges):

```
s1;
g1;
s2;
g2; (* we run out of gas here *)
s3;
g3;
```

We now optimise into this:
```
s1;
s2;
s3;
g1;
g2; (* We run out of gas here *)
g3;
```

Now the miner has to execute s1, s2 and s3 before running out of gas, whereas before it ran out of gas after executing only s1 and s2. This is unfair to the miner.

You may then say that this is not a legal optimisation because work that is "scheduled" to happen after an out-of-gas error must not be moved to before the out-of-gas-error. But then I would counter that no optimisation involving multiple statements would be legal. because every gas charge statement may cause an out-of-gas error, so we can't move statements around, i.e., gas charge statements prevent optimisations at statement level.

vaivaswatha commented:
> [snip]
> 
> > > My worry is that either compiler optimisations will be impossible (because gas charge statements prevent us from combining statements, e.g., through deforestation),
> > 
> > 
> > I agree that, to an extent, compiler optimizations may be impacted, but not made impossible. You can still, for example, combine statements. You'll have to be more smart about it. For example, do both computations first, and then both gas charges. If the gas charge is before a computation, do both charges before, and then both the computations (this is safe because, it would run out of gas anyway if it was supposed to, but now, without actually doing that computation).
> 
> Except that with this change you don't know (for statements) whether a gas charge is connected to the previous or the next statement. We knew that before, so we could just optimise away without worrying about gas, and the gas charge would follow the optimised code.
> 
> Also, it's not just about whether we run out of gas or not. It's also about the amount of work the miner needs to do before we run out of gas. Say that your program is as follows (`s` for statement, `g` for gas charges):
> 
> ```
> s1;
> g1;
> s2;
> g2; (* we run out of gas here *)
> s3;
> g3;
> ```
> 
> We now optimise into this:
> 
> ```
> s1;
> s2;
> s3;
> g1;
> g2; (* We run out of gas here *)
> g3;
> ```
> 
> Now the miner has to execute s1, s2 and s3 before running out of gas, whereas before it ran out of gas after executing only s1 and s2. This is unfair to the miner.
> 
> You may then say that this is not a legal optimisation because work that is "scheduled" to happen after an out-of-gas error must not be moved to before the out-of-gas-error. But then I would counter that no optimisation involving multiple statements would be legal. because every gas charge statement may cause an out-of-gas error, so we can't move statements around, i.e., gas charge statements prevent optimisations at statement level.

My point was that, you don't optimize it the way you said, but as below, assuming it isn't a "fetch value" kind of statement (because in that case, the gas cost is known only after the statement executes).

```
g1;
g2; (* We run out of gas here *)
g3;
s1;
s2;
s3;
```

In this case, it is unfair to neither the miner or the person executing the transition.

But in general, yes, optimizations need to be aware of gas, but I see no other way to ensure compatibility b/w gas charging in the interpreter and the VM.

jjcnn commented:
> > [snip]
[snip]
> My point was that, you don't optimize it the way you said, but as below, assuming it isn't a "fetch value" kind of statement (because in that case, the gas cost is known only after the statement executes).
> 
> ```
> g1;
> g2; (* We run out of gas here *)
> g3;
> s1;
> s2;
> s3;
> ```
> 
> In this case, it is unfair to neither the miner or the person executing the transition.

But that optimisation is also illegal, because `g1` and `g2` depend on values computed by `s1` and `s2`. So we can't move `gi` so that it happens before `si`, and we can't move `si` so that it happens before `g(i-1)`. Therefore, we can't swap gas charges with non-gas statements, and therefore we won't be able to combine statements. The only exception is when the sequence is `g1; s1; s2; g2`, but then `s2` is a `Load` statement (and `s1` is not), which as far as I can tell does not allow any optimisation to happen.

> 
> But in general, yes, optimizations need to be aware of gas, but I see no other way to ensure compatibility b/w gas charging in the interpreter and the VM.

It needs to be very aware of gas, to the point where no optimisations seem to be possible.

jjcnn reviewed (approved):
There is now only a discussion of the principle of gas charge nodes in relation to optimisations left as an open question, so we can merge this, and postpone the discussion until we actually want to do some optimisations.
